### PR TITLE
Workaround for MAX_PATH issues on Windows

### PR DESCRIPTION
**Pull request by crummel**
A few repos have extremely long file names for testing (AspNetCore, core-setup, and F#).  Unlike the Git command line, LibGit2Sharp does not attempt to mitigate MAX_PATH issues when checking out entire commits.  This change makes it so that if checking out the whole commit fails, we will instead check out individual files and just skip the ones that have paths that are too long.

There is some slight risk in that we will no longer catch if an actual source file cannot be checked out due to MAX_PATH.  We will emit a warning and the build will likely fail in that case.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs [hunk 1]
```diff
@@ -298,6 +288,60 @@ private static void CheckoutSubmodules(LibGit2Sharp.Repository repo, LibGit2Shar
             }
         }
 
+        private static void SafeCheckout(LibGit2Sharp.Repository repo, string commit, LibGit2Sharp.CheckoutOptions options, ILogger log)
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs]: How similar is this to the other SafeCheckout? Can we put the code in a common place if it's similar enough?

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 2]
```diff
@@ -467,6 +454,60 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
             }
         }
 
+        private static void SafeCheckout(LibGit2Sharp.Repository repo, string commit, LibGit2Sharp.CheckoutOptions options, ILogger log)
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Can you add a comment here that indicates the purpose and workflow for this method? There is a lot of try/catch logic in here which can be a bit hard to follow, so it would be good to comment something like:

```
// MAX_PATH is an issue in some repos. We can mitigate by doing X
// In this method, we attempt to checkout using the normal methodology
// if it fails in Y way, we try Z
// if it fails in W way, we try Q
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 3]
```diff
@@ -467,6 +454,70 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
...
+        {
+            try
+            {
+                log.LogDebug($"Trying safe checkout of {repo.Info.WorkingDirectory} at {commit}");
+                LibGit2Sharp.Commands.Checkout(repo, commit, options);
+            }
+            catch (Exception e) when (e is LibGit2Sharp.InvalidSpecificationException || e is LibGit2Sharp.NameConflictException || e is LibGit2Sharp.LibGit2SharpException)
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >catch (Exception e) when (e is LibGit2Sharp.InvalidSpecificationException || e is LibGit2Sharp.NameConflictException || e is LibGit2Sharp.LibGit2SharpException) [](start = 12, length = 160)

Nit: split into shorter lines

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 4]
```diff
@@ -467,6 +454,70 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
...
+            {
+                log.LogWarning($"Couldn't check out one or more files, possibly due to path length limitations ({e.ToString()}).  Attempting to checkout by individual files.");
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >log.LogWarning($"Couldn't check out one or more files, possibly due to path length limitations ({e.ToString()}).  Attempting to checkout by individual files."); [](start = 16, length = 160)

Nit: split into shorter lines

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 5]
```diff
@@ -467,6 +454,70 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
...
+                string resolvedReference = ParseReference(repo, commit, log);
+                if (resolvedReference != null)
+                {
+                    log.LogDebug($"Resolved {commit} to {resolvedReference}, attempting to check out");
+                    try
+                    {
+                        log.LogDebug($"Trying checkout of {repo.Info.WorkingDirectory} at {resolvedReference}");
+                        LibGit2Sharp.Commands.Checkout(repo, resolvedReference, options);
+                    }
+                    catch (Exception e) when (e is LibGit2Sharp.InvalidSpecificationException || e is LibGit2Sharp.NameConflictException || e is LibGit2Sharp.LibGit2SharpException)
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >catch (Exception e) when (e is LibGit2Sharp.InvalidSpecificationException || e is LibGit2Sharp.NameConflictException || e is LibGit2Sharp.LibGit2SharpException) [](start = 20, length = 160)

Nit: split into shorter lines

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 6]
```diff
@@ -467,6 +454,70 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
...
+                    {
+                        log.LogWarning($"Couldn't check out one or more files, possibly due to path length limitations ({e.ToString()}).  Attempting to checkout by individual files.");
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >log.LogWarning($"Couldn't check out one or more files, possibly due to path length limitations ({e.ToString()}).  Attempting to checkout by individual files."); [](start = 24, length = 160)

Nit: split into shorter lines

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 4 above)
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >LogWarning [](start = 20, length = 10)

Do we ever plan to use this flow in AzDo builds? If so, this warning might make the build to fail if `warnAsError` is set to true which is the case by default

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 7]
```diff
@@ -467,6 +454,70 @@ private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger
...
+        }
+
+        private static void SafeCheckoutByIndividualFiles(LibGit2Sharp.Repository repo, string commit, LibGit2Sharp.CheckoutOptions options, ILogger log)
+        {
+            log.LogDebug($"Beginning individual file checkout for {repo.Info.WorkingDirectory} at {commit}");
+            SafeCheckoutTreeByIndividualFiles(repo, repo.Lookup(commit).Peel<LibGit2Sharp.Tree>(), "", commit, options, log);
+
+        }
+
+        private static void SafeCheckoutTreeByIndividualFiles(LibGit2Sharp.Repository repo, LibGit2Sharp.Tree tree, string treePath, string commit, LibGit2Sharp.CheckoutOptions options, ILogger log)
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >rivate static void SafeCheckoutTreeByIndividualFiles(LibGit2Sharp.Repository repo, LibGit2Sharp.Tree tree, string treePath, string commit, LibGit2Sharp.CheckoutOptions options, ILogger log) [](start = 9, length = 189)

Nit: split into shorter lines

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs (hunk 1 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs]: Consolidated this in a helper class.

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 2 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Added a comment explaining.

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 3 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Fixed.

---
In reply to: [285641169](https://github.com/dotnet/arcade-services/pull/422#discussion_r285641169) [](ancestors = 285641169)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 4 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Changed as suggested.

---
In reply to: [285641238](https://github.com/dotnet/arcade-services/pull/422#discussion_r285641238) [](ancestors = 285641238)

Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: I'm currently using this in an AzDo build and it doesn't seem to be triggering it.  I'm not opposed to it doing so though - I think someone having to look at this and say it's okay that files aren't getting checked out is desirable behavior.

---
In reply to: [285644361](https://github.com/dotnet/arcade-services/pull/422#discussion_r285644361) [](ancestors = 285644361)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 5 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Changed as suggested.

---
In reply to: [285641356](https://github.com/dotnet/arcade-services/pull/422#discussion_r285641356) [](ancestors = 285641356)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 6 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Changed as suggested.

---
In reply to: [285641411](https://github.com/dotnet/arcade-services/pull/422#discussion_r285641411) [](ancestors = 285641411)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 7 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Changed as suggested.

---
In reply to: [285644577](https://github.com/dotnet/arcade-services/pull/422#discussion_r285644577) [](ancestors = 285644577)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 4 above)
Comment (jcagme) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: I totally agree, just wanted to point that since using warnings has hit us in the past and I remember we ended up switching to a info/debug log.

**Comment by crummel** [comment]
@mmitche Does this look good to merge?
